src/network/NetworkManagerSecondary.cc

```diff
@@ -96,86 +117,128 @@ bool NetworkManagerSecondary::OnControl(const private_msgs::PeerControl &_req,
   return true;
 }
 
-/////////////////////////////////////////////////
-void NetworkManagerSecondary::OnStep(
-    const private_msgs::SimulationStep &_msg)
+//////////////////////////////////////////////////
+void NetworkManagerSecondary::AsyncStepTask()
 {
-  IGN_PROFILE("NetworkManagerSecondary::OnStep");
-
-  // Throttle the number of step messages going to the debug output.
-  if (!_msg.stats().paused() && _msg.stats().iterations() % 1000 == 0)
-  {
-    igndbg << "Network iterations: " << _msg.stats().iterations()
-           << std::endl;
-  }
+  std::deque<private_msgs::SimulationStep> next_steps;
+  uint64_t maxIteration{0};
+  UpdateInfo lastUpdateInfo;
 
-  // Update affinities
-  for (int i = 0; i < _msg.affinity_size(); ++i)
+  while (!this->stopAsyncStepThread)
   {
-    const auto &affinityMsg = _msg.affinity(i);
-    const auto &entityId = affinityMsg.entity().id();
-
-    if (affinityMsg.secondary_prefix() == this->Namespace())
-    {
-      this->performers.insert(entityId);
+    IGN_PROFILE("NetworkManagerSecondary::AsyncStepTaskLoop");
+
+    if (next_steps.size()) {
+      // Ack message received
+      auto next_step = next_steps.front();
+      next_steps.pop_front();
+      auto info = convert<UpdateInfo>(next_step.stats());
+      // Max iteration is get from the primary message.
+      // lastUpdateInfo is updated with the primary provided message
+      // but `simTime` and `iterations` is not updated as the secondary
+      // may be "ahead".
+      // TODO(ivanpauno): Does realTime matter in the secondaries?
+      maxIteration = info.iterations + next_step.max_iterations();
+      uint64_t currentIteration = lastUpdateInfo.iterations;
+      auto simTime = lastUpdateInfo.simTime;
+      lastUpdateInfo = info;
+      lastUpdateInfo.iterations = currentIteration;
+      lastUpdateInfo.simTime = simTime;
+
+      // Throttle the number of step messages going to the debug output.
+      if (!info.paused && info.iterations % 1000 == 0)
+      {
+        igndbg << "Network iterations: " << info.iterations
+              << std::endl;
+      }
 
-      ignmsg << "Secondary [" << this->Namespace()
-             << "] assigned affinity to performer [" << entityId << "]."
-             << std::endl;
-    }
-    // If performer has been assigned to another secondary, remove it
-    else
-    {
-      auto parent =
-          this->dataPtr->ecm->Component<components::ParentEntity>(entityId);
-      this->dataPtr->ecm->RequestRemoveEntity(parent->Data());
-
-      if (this->performers.find(entityId) != this->performers.end())
+      // Update affinities
+      // TODO(ivanpauno):
+      // This needs to be redone based on if performers can "far apart",
```

Comment:
I think based on our current design, performer affinities should never change.  The plugins specific for a controller (aside from physics) will be executed locally on each secondary, and will not shift.

What will change is that the syncronization of physics will either be more strict or collocated if necessary.

Reply:
I think a bit more on this before our meeting on monday.

Reply:
This won't be addressed in the current PR